Lib/contextlib.py

```diff
@@ -164,7 +164,9 @@ def __exit__(self, typ, value, traceback):
                 # Suppress StopIteration *unless* it's the same exception that
                 # was passed to throw().  This prevents a StopIteration
                 # raised inside the "with" statement from being suppressed.
-                return exc is not value
+                result = exc is not value
+                del exc, value  # avoid cyclic reference
```

Comment:
You'll need to make the appropriate change in asynccontextmanager, and add a test

Reply:
I do not think that this is the right fix. Should not all local variables be cleared after leaving the function? If they are not, we have a larger issue.

Reply:
the traceback of the exception refers to the local variables

Reply:
I think the fix here is to wrap this whole `__exit__` in a:

```python
try:
    ...
finally:
    del value, traceback
```

Reply:
@graingert or maybe this should be done one lever higher? Do other `__exit__`s have the same problem?

I agree with @serhiy-storchaka, this fix is not correct.

Reply:
It would be pretty magical to automatically delete value and traceback from every `__exit__` and `__aexit__`

Reply:
Why this was not an issue in 3.11?

I agree that `finally` with `del`s is better solution, but we should look if there is more fundamental issue, which affects not only `contextmanager`.

Reply:
I don't entirely know why it was not an issue in Python 3.11, but concretely, if you add `print(self.__traceback__.tb_frame.f_back.f_locals)` in `__del__` method, in Python 3.12 it prints out `{ ... 'value': MyException() ... }` while in Python 3.11 it prints out `AttributeError: 'NoneType' object has no attribute 'f_locals'`.